### PR TITLE
feat(tailwind): Smaller bundle size

### LINKED ISSUE #1101: @react-email/tailwind Package dist size is too large
### Describe the Bug

I am trying to use react-email with TRPC and Next.js I ran a profiler on the output of `next build` with `@next/bundle-analyzer` It said the largest package was coming from `@react-email/tailwind/dist/index.mjs`

I am running Next.js with Open-Next deployed with SST on a Lambda in AWS. The larger the Node.js portion of the next.js code is, the longer it takes for the lambda to start. Currently it looks like `@react-email/tailwind` is 6.5MB 

At first glance, it looks like it's pulling in all of tailwindcss and postcss here: https://github.com/resend/react-email/blob/canary/packages/tailwind/src/utils/get-css-for-markup.ts#L1-L4

Is there a way we can make this package smaller? 

### Which package is affected (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue

https://github.com/resend/react-email/tree/canary/packages/tailwind

### To Reproduce

Run `next build` with [`@next/bundle-analyzer`](https://www.npmjs.com/package/@next/bundle-analyzer)

### Expected Behavior

The dist folder to be smaller.

### What's your node version? (if relevant)

_No response_

### LINKED ISSUE #1105: Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
### Describe the Bug

Hello, this is the one I was referring to to in [the other ticket](https://github.com/resend/react-email/issues/1054) @gabrielmfern,

I have a monorepo with an auth package, an email package and my main app that consumes both of them.
My desire is for users to login through next-auth and have next-auth send out an email with resend of a login link that is styled with react email.
I have been trying to have the next-auth middleware run on edge in order to make it faster.

The problem that I'm experiencing is that the project works well on dev, but when trying to build and deploy to vercel, it fails with 

```bash
./node_modules/.pnpm/@react-email+tailwind@0.0.14-canary.0_react@18.2.0/node_modules/@react-email/tailwind/dist/index.mjs
Dynamic Code Evaluation (e. g. 'eval', 'new Function', 'WebAssembly.compile') not allowed in Edge Runtime
Learn More: https://nextjs.org/docs/messages/edge-dynamic-code-evaluation

Import trace for requested module:
./node_modules/.pnpm/@react-email+tailwind@0.0.14-canary.0_react@18.2.0/node_modules/@react-email/tailwind/dist/index.mjs
./node_modules/.pnpm/@react-email+components@0.0.14-canary.0_@types+react@18.2.39_react@18.2.0/node_modules/@react-email/components/dist/index.mjs
./src/vercel-invite-user.tsx
./src/auth.ts
```
I've been using the canary versions of both @react-email/components & @react-email/render cause they contain fixes that I need to run on edge : https://github.com/resend/react-email/pull/1079

I'm not sure if it's a regression or a problem caused by a different library eg. NextAuth or Next itsself. 

It took me a while to create the reproduction because I have attempted: 

1) See if the problem occurs in a NextJS API route. It works without a problem ✅
2) See if the problem occurs when having the renderAsync function in a RSC. It works without a problem ✅
3) Have the function in the NextAuth configuration. It fails ❌

### Which package is affected (leave empty if unsure)

@react-email/components

### Link to the code that reproduces this issue

https://github.com/VaniaPopovic/react-email-repro

### To Reproduce

1) pnpm install 
2) pnpm run dev -> everything works as expected
3) pnpm run build it fails with the aformentioned error

### Expected Behavior

For it to render the email without a problem

### What's your node version? (if relevant)

20.6.1

### LINKED ISSUE #1111: Key is missing when using tailwind
### Describe the Bug

Due to the way the tailwind component works, it renders an array of children, which requires a `key`.

So the key needs to be appended to the element like so (note the key would be the index of the `map`:

```tsx
export function appendKeyToValidElement(element: ReactNode, key: number): ReactNode {
  if (isValidElement(element) && element.key === null) {
    return cloneElement(element, { key });
  }
  return element;
}
```

### Which package is affected (leave empty if unsure)

@react-email/tailwind

### Link to the code that reproduces this issue

https://github.com/resend/react-email

### To Reproduce

Start the demo and render the tailwind example. You will see the following warning in the console:

image

### Expected Behavior

React elements should rendered with a valid key.

### What's your node version? (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by gabrielmfern**
This changes the way we run Tailwind by manually calling out the internal functions they have
to generate the CSS based on the class names that was very inspired by how the Tailwind LSP 
server computes styles for the hover. 

This different approach allows us to write the code in a bit more readable manner than it was before, 
avoid rendering everything twice, and avoid bundling of node-specific code which forced 
us into having to bundle polyfills that would cause large dist sizes that are now avoided due to Vite's tree shaking. 
This also has the benefit of giving something like a 30% speed bump!

This PR also replaces #1119, which I'll close right now.

Closes #1101, closes #1111

Maybe closes #1105?

---

Edit February 2024: I also added an `automated-test-next-app` that is built using the Tailwind in the directory
in a way that it reproduces an actual environment to make sure we don't introduce any regressions.

### CODE REVIEW

**Comment by gabrielmfern** [comment]
Converting this into a draft because I need to investigate weather or not this causes issues with builds on Next and in other similar situations that would still require us to have polyfills.

**Comment by nicitaacom** [comment]
@gabrielmfern can you please request review somewhere here?
image


Also how to use your version of `react-email` I mean with this fix?

File: packages/tailwind/automated-test-next-app/package-lock.json
[the remark this answers is not in the document]
Comment (gabrielmfern) [reply packages/tailwind/automated-test-next-app/package-lock.json:1]: Will give it a try rn.

[the remark this answers is not in the document]
Comment (gabrielmfern) [reply packages/tailwind/automated-test-next-app/package-lock.json:1]: Done

**Comment by gabrielmfern** [comment]
Closing this in favor of #1383.

**Pull request closed without merging**